proposals/hybrid-rendering.md

```diff
@@ -0,0 +1,108 @@
+
+However, there are use-cases and reasons for wanting to have some dynamic pages in a static site.
+
+## Use-cases
+
+A few of the use-cases collected when talking to users who have requested this feature:
+
+- A marketing site for an agency that contains a contact form. Most of the site can be static and served via CDN, but the endpoint to serve the contact form needs to be dynamic so that it can store the contact information in a database and alert the admin.
+- A SaaS product where the API is the main product. Dynamic parts of pages are built with client components, API endpoints are the only server routes needed.
+- A content site such as a recipe site that allows users to mark their favorite recipes. API routes would be dynamic, as would any pages that display the dynamic list of favorites.
+
+# Goals
+
+- Allow default-static apps to have some pages that are dynamic.
+- Align with the current implementation and prevent an extra code-path that will be difficult to maintain.
+- Provide a better path when a site goes from static to server-rendered. Hybrid is a nice middleground.
+
+# Non-Goals
+
+- Any extra features outside of marking certain pages to not be prerendered.
+
+# Detailed Design
+
+The Astro config definition and types will need to be updated to allow the `'hybrid'` value for `output`.
+
+In `packages/astro/src/core/routing/manifest/create.ts` each route is set up to be `prerender: false` by default. This should be changed to be based on the `output` config option. If `'hybrid'` it should be interpreted as true, otherwise it should remain false
+
+In `packages/astro/src/vite-plugin-scanner/scan.ts` it currently throws for falsey values. Since in hybrid rendering users will set `export const prerender = false`, this code will need to be updated to allow the false value when the `output: 'hybrid'`.
+
+Additionally there are a few places in the codebase that assume if `output !== 'static'` that it is server mode. Those might need to be changed, depending on what they are doing with that information. For example, in static mode you cannot access the `Astro.url.searchParams`. However, because this happens before we load a page component we cannot know at the time if the route is prerendered or not, so we cannot continue to enforce this restriction in hybrid rendering. For this reason, and for simplicity, in hybrid rendering these restrictions are lifted for all routes.
+
+# Testing Strategy
+
+Prerendering is currently tested via fixture testing, due to the fact that the build artifacts is what changes. This same strategy will be used to test hybrid rendering as well, only testing for the opposite effect. 
+
+Likely we can use the same test fixtures, but only swap out the `output` when testing `'hybrid'`, which eliminates the need for new fixtures.
```

Comment:
In some cases we will need to create new test fixture, like tests in this PR https://github.com/withastro/astro/pull/7150
```suggestion
Likely we can use most of the prerendering test fixtures, and only swap out the `output` when testing `'hybrid'`, which reduces the need for new fixtures.
```